std/socket.d

```diff
@@ -289,56 +289,96 @@ bool wouldHaveBlocked() nothrow @nogc
     assert(rec == -1 && wouldHaveBlocked());
 }
 
+// Returns true iff getnameinfo is supported on this platform.
+version(Windows) private @property bool getnameinfoSupported()
```

Comment:
Why add `getaddrinfoSupported` and `getnameinfoSupported` and not just make `getnameInfoPointer` / `getaddrinfoPointer` lazily-initialized property functions?

Reply:
That's not new to this PR - old code handled the cases when the pointers were null and used workaround solutions.

So, old code:

* try to fetch the pointers at startup
* compare the pointers against null in code

New code:

* define xxxSupported and xxx, both doing initialization on first call
* use xxxSupported instead of comparing pointers to null

Reply:
Right, but, why not just keep comparing pointers to null?

Reply:
Because you need to call initialize before comparing them to null.

Reply:
Yes, so, why not turn the function pointer *variables* into function pointer returning *property functions*, in the same way that `getnameinfoSupported` are in the current version, but returning the pointer directly instead of a `bool`?

Reply:
That would be an alternative, but it seems more roundabout. To make sure I understand:

```D
@property auto our_getnameinfoPtr() {
    initialize;
    return &getNameInfoPtr;
}
... using code ...
if (our_getnameinfoPtr)
{
    ...
    our_getnameinfoPtr(... arguments ...);
}
```

Is this the idea? Mine was to not use pointers at all, but instead have the simple "if xxxSupported returns true, you may call xxx" convention. It seems yours could be used to avoid two calls to initialize by using local variables:

```d
... using code ...
auto gni = our_getnameinfoPtr;
if (gni)
{
    ...
    gni(... arguments ...);
}
```

Reply:
> That would be an alternative, but it seems more roundabout.

I don't see how that would be more roundabout. If anything, adding an additional pair of functions would better fit the description of "roundabout".

> It seems yours could be used to avoid two calls to initialize by using local variables:

I don't see the need to avoid performing two calls. The compiler ought to inline the function, at which point it can trivially conclude that under the `if` branch, the pointer cannot possibly be null, and eliminate the second comparison.

Reply:
Cool, yeah, two functions are better than four.